src/beautifiers/clang-format.coffee

```diff
@@ -0,0 +1,60 @@
+###
+Requires clang-format (https://clang.llvm.org)
+###
+
+"use strict"
+Beautifier = require('./beautifier')
+path = require('path')
+fs = require('fs')
+
+module.exports = class ClangFormat extends Beautifier
+
+  name: "clang-format"
+
+  options: {
+    "C++": true
+    "C": true
+    "Objective-C": true
+  }
+
+  ###
+    Dump contents to a given file
+  ###
+  dumpToFile: (name = "atom-beautify-dump", contents = "") ->
+    return new Promise((resolve, reject) =>
+      fs.open(name, "w", (err, fd) =>
+        @debug('dumpToFile', name, err, fd)
+        return reject(err) if err
+        fs.write(fd, contents, (err) ->
+          return reject(err) if err
+          fs.close(fd, (err) ->
+            return reject(err) if err
+            resolve(name)
+          )
+        )
+      )
+    )
+
+  beautify: (text, language, options) ->
+    return new @Promise((resolve, reject) ->
+      editor = atom.workspace.getActiveTextEditor()
```

Comment:
Thanks for your careful review. :+1:  But since clang-format uses a directory based approach to find the configuration file, we need to provide it with correct path. 

To be specific, to format file `/aa/bb/cc/dd/ee/ff.cpp`, clang-format will find in directories `[/aa/bb/cc/dd/ee, /aa/bb/cc/dd, /aa/bb/cc, /aa/bb, /aa, /]` in the apperance order for a file `.clang-format`. Project usually put this file in the project root and users would like to put a global `.clang-format` in their home directory. If we use tempFile, the file is usually created in `/tmp/xxx`, so no `.clang-format` can be found in the way, which causes clang-format uses default format style, which is only good for llvm (whose format is the default of clang-format).

To workaround the above problem, I request the editor path and constructed a temp file in the same directory as the original file, and delete it after use. This is why the code deliberately creates a 'dumpToFile' function.

If there are better approaches, I would like very much to explore it.